base/iterators.jl

```diff
@@ -152,15 +152,23 @@ end
 @inline done(r::Reverse{<:Enumerate}, state) = state[1] < 1
 
 """
-    Iterators.IndexValue(values, keys) <: AbstractDict{eltype(keys), eltype(values)}
+    Iterators.Pairs(values, keys) <: AbstractDict{eltype(keys), eltype(values)}
 
 Transforms an indexable container into an Dictionary-view of the same data.
 Modifying the key-space of the underlying data may invalidate this object.
 """
-struct IndexValue{K, V, I, A} <: AbstractDict{K, V}
+struct Pairs{K, V, I, A} <: AbstractDict{K, V}
     data::A
     itr::I
-    IndexValue(data::A, itr::I) where {A, I} = new{eltype(I), eltype(A), I, A}(data, itr)
+    Pairs(data::A, itr::I) where {A, I} = new{eltype(I), eltype(A), I, A}(data, itr)
+end
+
+function Base.showarg(io::IO, r::Pairs, toplevel)
+    if try typeof(pairs(r.data)) == typeof(r); catch; false; end
```

Comment:
Why is there a try/catch trying to hide on this line?

Reply:
This seems like a bad way to handle this.

Reply:
Seemed like `pairs` might throw?

Reply:
Well, this is an interactive-use function, so I didn't think it was too-terrible. The query is basically "Is this the type I'd get if I called `pairs` on this thing". Seemed like the most straightforward way to express that. We could also do something with `return_types`, but not sure that's better.

Reply:
I had a great reply typed out, then clicked the wrong button. C'est la vie, I suppose – second draft is always better, right?

I looked at how the other `showarg` functions are implemented, and noticed / realized that none of them care about whether the printed syntax is precise, just whether it is useful. That is to say, they only care if its a reasonable interpretation. None of them care to check if that interpretation is precise.

In conclusion, the implementation should simply be:
```julia
print(io, "pairs(")
showarg(io, r.data, false)
print(io, ")")
```

Reply:
And also define:
```julia
showarg(io, nt::NamedTuple, toplevel) = print(io, toplevel ? typeof(nt) : "::NamedTuple")
showarg(io, tt::Tuple, toplevel) = print(io, toplevel ? typeof(tt) : "::Tuple")
```

Reply:
> In conclusion, the implementation should simply be:

Well, The `Pairs` iterator allows different index sets, so e.g. `pairs(IndexCartesian(), A)` should not print as `pairs(A)`.

Reply:
I think the same could be said of the other `showarg` functions. Mostly, they'll be distinguished by the printed data. But we can just extend this to handle that case also.

Reply:
I don't think so. If we print a simple `pairs(` line, we should make sure it's correct. `We wouldn't print ReshapedArray as `reshape(::Array, rand(Int), rand(Int))`.